### ISSUE TITLE
MSGFPlusAdapter crashes

### ISSUE THREAD

**Issue by lars20070** [OpenMS/OpenMS#1764]
The MS-GF+ adapter crashes. Reported by Bassel Sabbagh (developer version, Windows). Also tested on developer version on OS X. MS-GF+ version 10089. Java installed. Any settings of `force` and `legacy_conversion`.

Error message: `Fatal error: Running MS-GF+ returned an error code. Does the MS-GF+ executable (.jar file) exist?`

**Comment by basselsabbagh** [comment]
Hello together, 

as Lars reported the MS-GF+ adapter crashes (see attached picture). When i tried to run the java file it seems it is not excutable. other java applications work without troubles.

hope you can help solving the problem 
regards
Bassel 

![error_msgfplus](https://cloud.githubusercontent.com/assets/16648406/12234345/a4bc4dc2-b86c-11e5-910a-95d76576af74.png)

**Comment by hendrikweisser** [comment]
Based on the "Could not reserve enough space" error: How much memory (RAM) does your computer have? Can you set the "java_memory" parameter to a smaller value (maybe 500) and try again?

**Comment by basselsabbagh** [comment]
my computer has 16 Gb RAM. I changed the Java memory setting and still getting the error.


________________________________

**Comment by hendrikweisser** [comment]
I see that the path that you pass to the "executable" parameter contains "MSGplus" as a folder name. Is there maybe just the "F" missing?

**Comment by lars20070** [comment]
@hendrikweisser The current head under Ubuntu is fine. With the current head under OS X I reproduce the MSGFPlusAdapter crash. (... and happy birthday!)

**Comment by basselsabbagh** [comment]
i added F to the folder name but the error still there!

**Comment by hendrikweisser** [comment]
What do you get when you run "java -jar C:\...\MSGFPlus.jar" (whatever the correct path is) on the command line?

**Comment by basselsabbagh** [comment]
unable to access jar file C:....

**Comment by hendrikweisser** [comment]
Interesting. Can you copy the .jar file to your user space and try to run it from there?

**Comment by basselsabbagh** [comment]
i put it on the desktop. it is working using the command line. but when i use it in TOPPAs it crashes

**Comment by lars20070** [comment]
Just talked with @basselsabbagh on the phone. He can now run MS-GF+ from the command line (java -jar MSGFPlus.jar) but the adapter still crashes. I see the same problem in OS X.

**Comment by hendrikweisser** [comment]
If you change the path in the adapter ("executable" parameter), do you get a different message before the crash?

**Comment by basselsabbagh** [comment]
no it is the same message

**Comment by hendrikweisser** [comment]
I'm at a loss.

Lars, if you can reproduce the problem on OS X, can you build an older version of the code and see up to what point it used to work?

**Comment by lars20070** [comment]
FYI, I installed Java and the current OpenMS nightly build on a new Windows machine. MSGFPlusAdapter runs just fine.

**Comment by lars20070** [comment]
In OS X within TOPPAS, the MSGFPlusAdapter crashes. In OS X from the command line, the MSGFPlusAdapter runs just fine.

**Comment by hendrikweisser** [comment]
I just tried the MSGFPlusAdapter with a current development build of OpenMS and didn't see any problems. I ran the tool tests from the command line ("ctest -R MSGFPlusAdapter") and they passed. Then I recreated the test in TOPPAS (workflow [here](https://gist.github.com/hendrikweisser/6837ef2db6619cd64054)) and that worked, too.

I put the "MSGFPlus.jar" file in my OpenMS "bin" directory. Maybe that makes a difference?

@basselsabbagh, @lars20070: Can you try whether my test workflow works on your machines?

**Comment by lars20070** [comment]
`ctest -R MSGFPlusAdapter` passes. Your TOPPAS workflow crashes, even with `MSGFPlus.jar` in OpenMS bin. As @ypriverol and me reported, MS-GF+ and `MSGFPlusAdapter` from the command line is fine, in TOPPAS we see the crashes.

**Comment by hendrikweisser** [comment]
@lars20070, which exact Java version do you use ("java -version")?

**Comment by lars20070** [comment]
`java version "1.7.0_45"
Java(TM) SE Runtime Environment (build 1.7.0_45-b18)
Java HotSpot(TM) 64-Bit Server VM (build 24.45-b08, mixed mode)`

**Comment by hendrikweisser** [comment]
Hm, that's not the same one as @ypriverol. You could try upgrading to the newest version, but I don't really think that it would solve the problem.

**Comment by basselsabbagh** [comment]
@hendrikweisser i tried it, still crashing! (i put the MSGFPlus.jar file inOpenMS bin directory)

**Comment by ypriverol** [comment]
@hendrikweisser I doubt the problem is in the java virtual machine. I guess should be in How you invoke it from the Adapter. Can be with security issues and Java preferences?

**Comment by lars20070** [comment]
@cbielow can you comment? How can `MSGFPlusAdapter` run from the command line but crash in `TOPPAS`?

**Comment by hendrikweisser** [comment]
@ypriverol: Java is called via Qt's `QProcess::execute`, but there's no difference in the call between the command line and TOPPAS.

**Comment by cbielow** [comment]
There should be no difference between a call in TOPPAS vs. command line.

However, TOPPAS working directory depends on how you start it. If you run it from command line, the WD is identical to the prompt. If you double-click a TOPPAS file, the path of that file is the WD.
I have no idea how MacOS does it, i.e, WD is the package root, or the executable directory or whatever... Depending on that, something might break if MSGF+ itself tries to write to its WD (which it will inherit from TOPPAS).

Also, what is the exact error message and on what system (Windows, MacOS)... this thread is somewhat confusing ....

**Comment by lars20070** [comment]
@cbielow The error message is always the same:  `Fatal error: Running MS-GF+ returned an error code. Does the MS-GF+ executable (.jar file) exist?` MS-GF+ cannot be started.

@basselsabbagh Encounters the problem in Windows. @ypriverol and me under OS X.

**Comment by hendrikweisser** [comment]
> The error message is always the same:

Judging from the screenshot that @basselsabbagh posted above, there is more to the error message at least in some cases: https://cloud.githubusercontent.com/assets/16648406/12234345/a4bc4dc2-b86c-11e5-910a-95d76576af74.png

**Comment by cbielow** [comment]
I guess the question is what the actual output of java is (if any). But that might require changing the code a little. Or take the debug steps I described. However, Having better instri sic debug info would be better also for future problems. ..
-- 
Sent via mobile phone.

Am 21. Januar 2016 17:31:59 MEZ, schrieb Lars Nilse <notifications@github.com>:

**Comment by hroest** [comment]
from the screenshot it rather seems like there is an issue with memory allocation:
```
Could not reserve enough space for xxx object heap
```
where it tries to allocate 3.5 GB of memory. Is this expected behavior? Is this how much it allocates from the commandline? Could it be that java uses different environmental variables when started from the command line then when started from a GUI? Can people run the code *without* GUI but using the MSGFPlusAdapter only from the commandline? 

It seems that the adapter by default allocates 3.5 GB of memory when run through the TOPP tool, see https://github.com/OpenMS/OpenMS/blob/develop/src/topp/MSGFPlusAdapter.cpp#L204 so to actually compare the command line execution, it would be important to give java the exact same parameters as in the tool. To debug this it would be great to (i) get the exact command that TOPPAS executes (e.g. `MSGFPlusAdapter ...`) and the exact code that MSGFPlusAdapter executes (e.g. `java -XX:MaxPermSize= ... `)

**Comment by hendrikweisser** [comment]
> where it tries to allocate 3.5 GB of memory. Is this expected behavior? Is this how much it allocates from the commandline?

Yes, this is expected and also happens from the command line. 3.5 GB is the default used by MS-GF+. The value can be changed using the "java_memory" parameter of the adapter, which just gets passed to the Java call ("-Xmx").

**Comment by hroest** [comment]
yes, but when you run it from the commandline, do you use the exact same
parameters?

**Comment by lars20070** [comment]
As @hendrikweisser pointed out, the crash from @basselsabbagh in Windows and the crash from @ypriverol under OS X might be two different issues.

@basselsabbagh cannot run MS-GF+ from the command line.
`java -Xmx3500M -jar MSGFPlus.jar -s test.mzML -d database.fasta -o test.mzid`
results in
` unable to access jarfile MSGFPlus.jar`
although `jar`, `mzML` and `fasta` are in the same folder. OpenMS and the adapter are not the problem.

**Comment by ypriverol** [comment]
Hi All:
 Didn't say anything because I'm trying to setup everything here to been able to compile from scratch and debug.

**Comment by hendrikweisser** [comment]
I tried it again on my MacBook after OS X was updated from 10.8.5 to 10.10.5. It still works for me.

**Comment by hroest** [comment]
@lars20070 did you set the correct permissions for the file?

**Comment by lars20070** [comment]
Looks fine to me, @hroest.
```
-rwxr-xr-x@ 1 lars  staff  21488983 Jul 16  2014 MSGFPlus.jar
```

**Comment by cbielow** [comment]
It seems there were two separate issues here:
 - Java executable not found (since MacOSX app bundles break the PATH) (solved by #2066)
 - Java running out of memory (solvable by increasing the heap memory -- or maybe not if you are on a 32bit Operating System, but thats another story)

Can we close this?

**Comment by lars20070** [comment]
IMO, yes. Great job @cbielow.
